### PR TITLE
Typosquatting optimization: add caching

### PR DESCRIPTION
**Pull request by zhhyu**
Add the cache of checklist to speed up.
Fix: https://github.com/NuGet/Engineering/issues/1809

### CODE REVIEW

File: src/NuGetGallery/App_Start/DefaultDependenciesModule.cs [hunk 1]
```diff
@@ -331,7 +331,7 @@ protected override void Load(ContainerBuilder builder)
             builder.RegisterType<TyposquattingService>()
                 .AsSelf()
                 .As<ITyposquattingService>()
-                .InstancePerLifetimeScope();
+                .SingleInstance();
```
[the remark this answers is not in the document]
Comment (zhhyu) [reply src/NuGetGallery/App_Start/DefaultDependenciesModule.cs]: Change back to lifetime scope and add another singleton instance for caching.

File: src/NuGetGallery/Services/TyposquattingService.cs [hunk 2]
```diff
@@ -19,8 +19,12 @@ public class TyposquattingService : ITyposquattingService
             new ThresholdInfo (lowerBound: 30, upperBound: 50, threshold: 1),
             new ThresholdInfo (lowerBound: 50, upperBound: 129, threshold: 2)
         };
+        private static readonly object Locker = new object();
```
[the remark this answers is not in the document]
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingService.cs]: If it's not static, multiple requests will have different locks because it's life time scope rather than singleton. Is that right?

File: src/NuGetGallery/Services/TyposquattingService.cs [hunk 3]
```diff
@@ -19,72 +19,94 @@ public class TyposquattingService : ITyposquattingService
             new ThresholdInfo (lowerBound: 30, upperBound: 50, threshold: 1),
             new ThresholdInfo (lowerBound: 50, upperBound: 129, threshold: 2)
         };
-
+        private static readonly object Locker = new object();
```
[the remark this answers is not in the document]
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingService.cs]: Move to cache class now.

File: src/NuGetGallery/Services/TyposquattingService.cs [hunk 4]
```diff
@@ -19,72 +19,93 @@ public class TyposquattingService : ITyposquattingService
...
+                                    TyposquattingCheckListCache typosquattingCheckListCache)
         {
             _contentObjectService = contentObjectService ?? throw new ArgumentNullException(nameof(contentObjectService));
             _packageService = packageService ?? throw new ArgumentNullException(nameof(packageService));
             _reservedNamespaceService = reservedNamespaceService ?? throw new ArgumentNullException(nameof(reservedNamespaceService));
             _telemetryService = telemetryService ?? throw new ArgumentNullException(nameof(telemetryService));
+            _typosquattingCheckListCache = typosquattingCheckListCache ?? throw new ArgumentNullException(nameof(typosquattingCheckListCache));
 
             TyposquattingCheckListLength = _contentObjectService.TyposquattingConfiguration.PackageIdChecklistLength;
+            _typosquattingCheckListCache.DefaultExpireTime = TimeSpan.FromDays(1);
```
Comment (joelverhagen) [inline src/NuGetGallery/Services/TyposquattingService.cs]: These should be initialized on the singleton, not per `TyposquattingService`. With this code, every time we create a new `TyposquattingService`, the last refresh time will be moved forward. Not sure, but I think this will make the cache never expire.

Comment (joelverhagen) [reply src/NuGetGallery/Services/TyposquattingService.cs]: I think we should have a unit test that initializes two `TyposquattingService` and verifies that one initializes the cache and the next uses the cached value without hitting the DB again.

File: src/NuGetGallery/Services/TyposquattingService.cs [hunk 5]
```diff
@@ -19,72 +19,93 @@ public class TyposquattingService : ITyposquattingService
...
+            }
+            var packageIdsCheckList = _typosquattingCheckListCache.Cache;
 
             var algorithmProcessingStopwatch = Stopwatch.StartNew();
             var threshold = GetThreshold(uploadedPackageId);
             var normalizedUploadedPackageId = TyposquattingStringNormalization.NormalizeString(uploadedPackageId);
-
             var collisionIds = new ConcurrentBag<string>();
-            Parallel.ForEach(packagesCheckList, (packageId, loopState) =>
+            Parallel.ForEach(packageIdsCheckList, (packageId, loopState) =>
```
Comment (agr) [inline src/NuGetGallery/Services/TyposquattingService.cs]: What happens if `packageIdsCheckList` content changes (cache update occurs in concurrent thread)?

File: src/NuGetGallery/Services/TyposquattingCheckListCache.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+using System.Collections.Generic;
+
+namespace NuGetGallery
+{
+    public class TyposquattingCheckListCache
+    {
+        public object Locker { get; } = new object();
+        public List<string> Cache { get; set; }
+        public DateTime LastRefreshTime { get; set; }
+        public TimeSpan DefaultExpireTime { get; set; }
```
Comment (agr) [inline src/NuGetGallery/Services/TyposquattingCheckListCache.cs]: I'd prefer to see the cache object a bit more self-contained and protected from accidental changes. Right now it exposes all it's guts to anyone who has a reference to it allowing to change the cached content in an uncontrollable fashion, or change its expiration status, or even worse acquire update lock and keep it forever, preventing actual updates.

I'd suggest moving the refresh code into the cache object, make `Locker` private and make `Cache` `IReadOnlyCollection<string>` for an outside observer, so it cannot be modified from outside of the class.
Similarly, `LastRefreshTime` and `DefaultExpireTime` should not have public setters. Probably they should not be public at all, since for the user of that class it only matters if it is expired or not.

Additionally, I would extract interface and put (and consume) this object as an interface, not as a concrete implementation. This will reduce coupling and make testing easier.

File: src/NuGetGallery/Services/TyposquattingService.cs [hunk 7]
```diff
@@ -19,72 +19,93 @@ public class TyposquattingService : ITyposquattingService
...
+                {
+                    if (_typosquattingCheckListCache.Cache == null || IsCheckListCacheExpired())
+                    {
+                        checklistRetrievalStopwatch = Stopwatch.StartNew();
+                        _typosquattingCheckListCache.Cache = _packageService.GetAllPackageRegistrations()
+                            .OrderByDescending(pr => pr.IsVerified)
+                            .ThenByDescending(pr => pr.DownloadCount)
+                            .Select(pr => pr.Id)
+                            .Take(TyposquattingCheckListLength)
+                            .ToList();
```
Comment (agr) [inline src/NuGetGallery/Services/TyposquattingService.cs]: `ToListAsync()`?

File: src/NuGetGallery/Services/TyposquattingCheckListCache.cs (hunk 6 above)
Comment (agr) [reply src/NuGetGallery/Services/TyposquattingCheckListCache.cs]: You might consider having the following interface:

```
public interface ISomethingSomethingCache
{
   string IReadOnlyCollection<string> Cache;
}
```
(or getter method if you want it `async`)
and the implementation would internally check for the expiration and refresh if needed. This would completely seal the implementation from the outside observer.

File: src/NuGetGallery/Services/TyposquattingService.cs (hunk 7 above)
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingService.cs]: We keep the typo-squatting check as a sync method.

File: src/NuGetGallery/Services/TyposquattingService.cs [hunk 8]
```diff
@@ -19,72 +19,93 @@ public class TyposquattingService : ITyposquattingService
...
             algorithmProcessingStopwatch.Stop();
 
-            var totalTime = checklistRetrievalStopwatch.Elapsed.Add(algorithmProcessingStopwatch.Elapsed);
-            _telemetryService.TrackMetricForTyposquattingChecklistRetrievalTime(uploadedPackageId, checklistRetrievalStopwatch.Elapsed);
             _telemetryService.TrackMetricForTyposquattingAlgorithmProcessingTime(uploadedPackageId, algorithmProcessingStopwatch.Elapsed);
+            var totalTime = algorithmProcessingStopwatch.Elapsed;
+
+            if (checklistRetrievalStopwatch != null)
+            {
+                _telemetryService.TrackMetricForTyposquattingChecklistRetrievalTime(uploadedPackageId, checklistRetrievalStopwatch.Elapsed);
```
Comment (agr) [inline src/NuGetGallery/Services/TyposquattingService.cs]: Why not report the metric as soon as we know its value? If anything throws between the cache refresh code and this line, we'll lose the data point.

File: tests/NuGetGallery.Facts/Services/TyposquattingServiceFacts.cs [hunk 9]
```diff
@@ -39,36 +35,64 @@ public class TyposquattingServiceFacts
...
+            {
+                telemetryService = new Mock<ITelemetryService>();
+            }
+
+            if (typosquattingCheckListCache == null)
+            {
+                typosquattingCheckListCache = new TyposquattingCheckListCache();
+            }
+
+            return new TyposquattingService(contentObjectService.Object, packageService.Object, reservedNamespaceService.Object, telemetryService.Object, typosquattingCheckListCache); ;
```
Comment (agr) [inline tests/NuGetGallery.Facts/Services/TyposquattingServiceFacts.cs]: Extra semicolon at the end of the line.

File: src/NuGetGallery/Services/TyposquattingService.cs (hunk 4 above)
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingService.cs]: Decouple to a new interface.

File: src/NuGetGallery/Services/TyposquattingService.cs (hunk 5 above)
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingService.cs]: I think the lock will handle this.

File: src/NuGetGallery/Services/TyposquattingCheckListCache.cs (hunk 6 above)
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingCheckListCache.cs]: Decouple to a new interface.

File: src/NuGetGallery/Services/TyposquattingService.cs (hunk 8 above)
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingService.cs]: Add to the telemetry immediately after getting.

File: tests/NuGetGallery.Facts/Services/TyposquattingServiceFacts.cs (hunk 9 above)
Comment (zhhyu) [reply tests/NuGetGallery.Facts/Services/TyposquattingServiceFacts.cs]: Delete it now.

File: src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        private static DateTime LastRefreshTime;
+        private static TimeSpan DefaultExpireTime;
+
+        private static int TyposquattingCheckListLength;
+
+        public TyposquattingCheckListCacheService()
+        {
+            TyposquattingCheckListLength = -1;
+            DefaultExpireTime = TimeSpan.FromDays(1);
+            LastRefreshTime = DateTime.UtcNow;
```
Comment (skofman1) [inline src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: this is not a good initialization value since it gives the impression the cache is up to date. Use DateTime.MinValue

File: src/NuGetGallery/Services/TyposquattingService.cs (hunk 5 above)
Comment (agr) [reply src/NuGetGallery/Services/TyposquattingService.cs]: This `ForEach` happens outside of the lock as far as I can see.

Comment (agr) [reply src/NuGetGallery/Services/TyposquattingService.cs]: I think it's fine though, since cache update does not mutate existing list, but generate a new one. So it should be fine.

File: src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        }
+
+        public IReadOnlyCollection<string> GetTyposquattingCheckList(int checkListLength, IPackageService packageService)
+        {
+            if (checkListLength < 0)
+            {
+                throw new ArgumentNullException(nameof(checkListLength));
```
Comment (agr) [inline src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: `ArgumentOutOfRangeException`

File: src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+            }
```
Comment (agr) [inline src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: Need to check `packageService` for `null`

File: src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+{
+    public class TyposquattingCheckListCacheService : ITyposquattingCheckListCacheService
+    {
+        private static readonly object Locker = new object();
+
+        private static List<string> Cache;
+        private static DateTime LastRefreshTime;
+        private static TimeSpan DefaultExpireTime;
+
+        private static int TyposquattingCheckListLength;
```
Comment (agr) [inline src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: Isn't it always the same as `Cache.Length`? Can we remove it?

File: src/NuGetGallery/App_Start/DefaultDependenciesModule.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (agr) [reply src/NuGetGallery/App_Start/DefaultDependenciesModule.cs]: Alternative solution would be instead of holding on a reference to an instance of `IPackageService`, request a factory producing instances of that service from DI: change `IPackageService packageService` to `Func<IPackageService> packageServiceFactory` in constructor. Then in `GetTyposquattingCheckList` method, you'd need not to pass an instance as an argument, but generate one as needed. Also it would not require `static` members in the cache class, which in my opinion is a cleaner approach.

File: src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs (hunk 13 above)
Comment (shishirx34) [reply src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: This value makes no sense in this context. You are always passing the `checklistLength` to the `GetTyposquattingCheckList` method. Why set this as a field on this service?

Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: Even though we always pass the checklistLength, we want to refresh the cache when length changes like from 5000 to 10000. If we do not keep this value, how should we know whether this length is changed or not?

Comment (shishirx34) [reply src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: you can remove this field.

File: src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs (hunk 10 above)
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: Change to MinValue now.

File: src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs (hunk 11 above)
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: Done

File: src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs (hunk 12 above)
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: Add that check now.

File: src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs (hunk 13 above)
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: We decide to keep the current approach now, because other solutions may lead to more complexity.

File: src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+        public TyposquattingCheckListCacheService()
+        {
+            TyposquattingCheckListLength = -1;
+            DefaultExpireTime = TimeSpan.FromDays(1);
+            LastRefreshTime = DateTime.MinValue;
+        }
+
+        public IReadOnlyCollection<string> GetTyposquattingCheckList(int checkListLength, IPackageService packageService)
+        {
+            if (checkListLength < 0)
```
Comment (agr) [inline src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: Should 0 be invalid, too?

File: src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+            {
+                throw new ArgumentOutOfRangeException(nameof(checkListLength));
```
Comment (agr) [inline src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: Please, provide an exception message stating the allowed range.

File: src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs (hunk 14 above)
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: I think we may allow to set check list length as 0.

File: src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace NuGetGallery
+{
+    public class TyposquattingCheckListCacheService : ITyposquattingCheckListCacheService
+    {
+        private static readonly object Locker = new object();
```
Comment (joelverhagen) [inline src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: These should not be `static`.

File: src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+        private static List<string> Cache;
+        private static DateTime LastRefreshTime;
+        private static TimeSpan DefaultExpireTime;
+
+        private static int TyposquattingCheckListLength;
+
+        public TyposquattingCheckListCacheService()
+        {
+            TyposquattingCheckListLength = -1;
+            DefaultExpireTime = TimeSpan.FromDays(1);
```
Comment (joelverhagen) [inline src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: This should probably be configurable.

File: src/NuGetGallery/Services/TyposquattingService.cs [hunk 18]
```diff
@@ -19,20 +19,25 @@ public class TyposquattingService : ITyposquattingService
...
+                                    ITelemetryService telemetryService,
+                                    ITyposquattingCheckListCacheService typosquattingCheckListCacheService)
         {
             _contentObjectService = contentObjectService ?? throw new ArgumentNullException(nameof(contentObjectService));
             _packageService = packageService ?? throw new ArgumentNullException(nameof(packageService));
             _reservedNamespaceService = reservedNamespaceService ?? throw new ArgumentNullException(nameof(reservedNamespaceService));
             _telemetryService = telemetryService ?? throw new ArgumentNullException(nameof(telemetryService));
+            _typosquattingCheckListCacheService = typosquattingCheckListCacheService ?? throw new ArgumentNullException(nameof(typosquattingCheckListCacheService));
 
             TyposquattingCheckListLength = _contentObjectService.TyposquattingConfiguration.PackageIdChecklistLength;
```
Comment (joelverhagen) [inline src/NuGetGallery/Services/TyposquattingService.cs]: nit: this does not need to be a field.

File: tests/NuGetGallery.Facts/Services/TyposquattingCheckListCacheServiceFacts.cs [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+            var newService = new TyposquattingCheckListCacheService();
+
+            // Act
+            int tasksNum = 3;
+            Task[] tasks = new Task[tasksNum];
+            for (int i = 0; i < tasksNum; i++)
+            {
+                tasks[i] = Task.Factory.StartNew(() =>
+                {
+                    newService.GetTyposquattingCheckList(_packageIds.Count, mockPackageService.Object);
```
Comment (joelverhagen) [inline tests/NuGetGallery.Facts/Services/TyposquattingCheckListCacheServiceFacts.cs]: per task, maybe execute this twice, so we have parallel _and_ sequential

File: tests/NuGetGallery.Facts/Services/TyposquattingCheckListCacheServiceFacts.cs [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Moq;
+using NuGet.Services.Entities;
+using Xunit;
+
+namespace NuGetGallery
+{
+    public class TyposquattingCheckListCacheServiceFacts
```
Comment (joelverhagen) [inline tests/NuGetGallery.Facts/Services/TyposquattingCheckListCacheServiceFacts.cs]: Make tests that check the various cache invalidation scenarios
 
- cache is too old
- cache has wrong length

File: src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs (hunk 15 above)
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: Add the message now.

File: src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs (hunk 16 above)
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: Delete them.

File: src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs (hunk 17 above)
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: Add the expire time parameter in configuration.

File: src/NuGetGallery/Services/TyposquattingService.cs (hunk 18 above)
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingService.cs]: Delete the field.

File: tests/NuGetGallery.Facts/Services/TyposquattingCheckListCacheServiceFacts.cs (hunk 19 above)
Comment (zhhyu) [reply tests/NuGetGallery.Facts/Services/TyposquattingCheckListCacheServiceFacts.cs]: Execute twice now with sequential test.

File: tests/NuGetGallery.Facts/Services/TyposquattingCheckListCacheServiceFacts.cs (hunk 20 above)
Comment (zhhyu) [reply tests/NuGetGallery.Facts/Services/TyposquattingCheckListCacheServiceFacts.cs]: Add these two tests for coverage.

File: src/NuGetGallery/Services/TyposquattingService.cs [hunk 21]
```diff
@@ -20,71 +20,68 @@ public class TyposquattingService : ITyposquattingService
...
-                .OrderByDescending(pr => pr.IsVerified)
-                .ThenByDescending(pr => pr.DownloadCount)
-                .Select(pr => pr.Id)
-                .Take(TyposquattingCheckListLength)
-                .ToList();
+            var packageIdsCheckList = _typosquattingCheckListCacheService.GetTyposquattingCheckList(checkListLength, checkListExpireTimeInHours, _packageService);
             checklistRetrievalStopwatch.Stop();
 
+            _telemetryService.TrackMetricForTyposquattingChecklistRetrievalTime(uploadedPackageId, checklistRetrievalStopwatch.Elapsed);
+            var totalTime = checklistRetrievalStopwatch.Elapsed;
```
Comment (joelverhagen) [inline src/NuGetGallery/Services/TyposquattingService.cs]: >totalTime [](start = 16, length = 9)

Why not measure total time directly? It seems error prone to add two timestamps together. Imagine if we add more business logic between the time that stopwatches are running... then there would be unaccounted for time.

File: src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+            if (checkListLength < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(checkListLength), "Negative values are not supported.");
+            }
+            if (packageService == null)
+            {
+                throw new ArgumentNullException(nameof(packageService));
+            }
+
+            if (Cache == null || checkListLength != TyposquattingCheckListLength || IsCheckListCacheExpired(checkListExpireTimeInHours))
```
Comment (joelverhagen) [inline src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: >Cache == null || checkListLength != TyposquattingCheckListLength || IsCheckListCacheExpired(checkListExpireTimeInHours) [](start = 16, length = 119)

These three conditions could be grouped into a method called `ShouldCacheBeUpdated` then you can call it in two places instead of duplicating the three conditions.

File: src/NuGetGallery/Services/TyposquattingService.cs (hunk 21 above)
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingService.cs]: Yes, that's true. Change it now.

File: src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs (hunk 22 above)
Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: Move to a function now.

File: src/NuGetGallery/Services/ITyposquattingCheckListCacheService.cs [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+    /// /// </summary>
+    public interface ITyposquattingCheckListCacheService
+    {
+        /// <summary>
+        /// The function is used to get the checklist from the cache for typosquatting
+        /// </summary>
+        /// <param name="checkListLength">The length of the checklist for typosquatting</param>
+        /// <param name="checkListExpireTimeInHours">The expire time for checklist caching</param>
+        /// <param name="packageService">The package service for checklist retrieval from database</param>
+        IReadOnlyCollection<string> GetTyposquattingCheckList(int checkListLength, double checkListExpireTimeInHours, IPackageService packageService);
```
Comment (shishirx34) [inline src/NuGetGallery/Services/ITyposquattingCheckListCacheService.cs]: the expiry time shouldn't be in unit of measurement, it should be `TimeSpan`

File: src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+
+        private int TyposquattingCheckListConfiguredLength;
+
+        public TyposquattingCheckListCacheService()
+        {
+            TyposquattingCheckListConfiguredLength = -1;
+            LastRefreshTime = DateTime.MinValue;
+        }
+
+        public IReadOnlyCollection<string> GetTyposquattingCheckList(int checkListConfiguredLength, double checkListExpireTimeInHours, IPackageService packageService)
```
Comment (shishirx34) [inline src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: don't pass in hours here, pass `TimeSpan`, the caller should determine if it should be hours or mins

File: src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+        }
+
+        private bool ShouldCacheBeUpdated(int checkListConfiguredLength, double checkListExpireTimeInHours)
+        {
+            return Cache == null || checkListConfiguredLength != TyposquattingCheckListConfiguredLength || IsCheckListCacheExpired(checkListExpireTimeInHours);
+        }
+
+        private bool IsCheckListCacheExpired(double checkListExpireTimeInHours)
+        {
+            return DateTime.UtcNow >= LastRefreshTime.Add(TimeSpan.FromHours(checkListExpireTimeInHours));
```
Comment (shishirx34) [inline src/NuGetGallery/Services/TyposquattingCheckListCacheService.cs]: no `FromHours` here.

File: src/NuGetGallery/Services/TyposquattingService.cs [hunk 26]
```diff
@@ -20,86 +20,84 @@ public class TyposquattingService : ITyposquattingService
...
             _telemetryService = telemetryService ?? throw new ArgumentNullException(nameof(telemetryService));
-
-            TyposquattingCheckListLength = _contentObjectService.TyposquattingConfiguration.PackageIdChecklistLength;
+            _typosquattingCheckListCacheService = typosquattingCheckListCacheService ?? throw new ArgumentNullException(nameof(typosquattingCheckListCacheService));
         }
 
         public bool IsUploadedPackageIdTyposquatting(string uploadedPackageId, User uploadedPackageOwner, out List<string> typosquattingCheckCollisionIds)
         {
+            var checkListConfiguredLength = _contentObjectService.TyposquattingConfiguration.PackageIdChecklistLength;
+            var checkListExpireTimeInHours = _contentObjectService.TyposquattingConfiguration.PackageIdChecklistCacheExpireTimeInHours;
```
Comment (shishirx34) [inline src/NuGetGallery/Services/TyposquattingService.cs]: convert to timespan here

File: src/NuGetGallery/Services/TyposquattingService.cs [hunk 27]
```diff
@@ -20,86 +20,84 @@ public class TyposquattingService : ITyposquattingService
...
+            var totalTimeStopwatch = Stopwatch.StartNew();
             var checklistRetrievalStopwatch = Stopwatch.StartNew();
-            var packageRegistrations = _packageService.GetAllPackageRegistrations();
-            var packagesCheckList = packageRegistrations
-                .OrderByDescending(pr => pr.IsVerified)
-                .ThenByDescending(pr => pr.DownloadCount)
-                .Select(pr => pr.Id)
-                .Take(TyposquattingCheckListLength)
-                .ToList();
+            var packageIdsCheckList = _typosquattingCheckListCacheService.GetTyposquattingCheckList(checkListConfiguredLength, checkListExpireTimeInHours, _packageService);
```
Comment (shishirx34) [inline src/NuGetGallery/Services/TyposquattingService.cs]: pass the above timespan

File: src/NuGetGallery/Services/TyposquattingService.cs [hunk 28]
```diff
@@ -20,86 +20,84 @@ public class TyposquattingService : ITyposquattingService
...
+                totalTimeStopwatch.Stop();
                 _telemetryService.TrackMetricForTyposquattingCheckResultAndTotalTime(
                     uploadedPackageId,
-                    totalTime,
+                    totalTimeStopwatch.Elapsed,
                     wasUploadBlocked,
                     typosquattingCheckCollisionIds,
-                    TyposquattingCheckListLength);
+                    packageIdsCheckList.Count,
+                    checkListExpireTimeInHours);
```
Comment (shishirx34) [inline src/NuGetGallery/Services/TyposquattingService.cs]: what do you get from passing the configured hours here?

Comment (zhhyu) [reply src/NuGetGallery/Services/TyposquattingService.cs]: Double check the cache expiration time we set in the configuration file.

File: tests/NuGetGallery.Facts/Services/TyposquattingCheckListCacheServiceFacts.cs [hunk 29]
```diff
@@ -0,0 +1,130 @@
...
+                new PackageRegistration()
+                {
+                    Id = _packageIds[i],
+                    DownloadCount = new Random().Next(0, 10000),
+                    IsVerified = true,
+                    Owners = new List<User> { new User() { Username = string.Format("owner{0}", i + 1), Key = i + 1 } }
+                }).AsQueryable();
+
+        [Fact]
+        public void CheckTyposquattingCheckListCache()
```
Comment (skofman1) [inline tests/NuGetGallery.Facts/Services/TyposquattingCheckListCacheServiceFacts.cs]: When reading UT code, you should be able to understand what the test checks. Consider using the following name: GetTyposquattingCheckListIsThreadSafe or
WhenGettingCheckListFromMultipleThreadsItIsInitializedOnce

File: tests/NuGetGallery.Facts/Services/TyposquattingCheckListCacheServiceFacts.cs [hunk 30]
```diff
@@ -0,0 +1,130 @@
...
+            // Assert
+            mockPackageService.Verify(
+               x => x.GetAllPackageRegistrations(),
+               Times.Exactly(2));
+        }
+
+        [Fact]
+        public void CheckRefreshedCheckListLengthNotAllowed()
+        {
+            // Arrange
```
Comment (skofman1) [inline tests/NuGetGallery.Facts/Services/TyposquattingCheckListCacheServiceFacts.cs]: add tests for input verification (null parameters, etc)

Comment (zhhyu) [reply tests/NuGetGallery.Facts/Services/TyposquattingCheckListCacheServiceFacts.cs]: Add some tests for cache.

File: tests/NuGetGallery.Facts/Services/TyposquattingCheckListCacheServiceFacts.cs (hunk 29 above)
Comment (zhhyu) [reply tests/NuGetGallery.Facts/Services/TyposquattingCheckListCacheServiceFacts.cs]: Change the names for cache tests.
